traffic light
either
south-north
east-west

right of way rules
green, turn left, if no coming straight or coming right
red, turn right, if no going left


In [ ]:
always at intersection
next state
stop, left, right, straight
time decreases for each action taken

rewards
1. complete trip - large
2. obey traffic rules - small

punishment
1. incorrect action - small
2. violate traffic rule/cause an accident - large




### QUESTION 1
Observe what you see with the agent's behavior as it takes random actions. Does the smartcab eventually make it to the destination? Are there any other interesting observations to note?

In [9]:
result = [60, 68, 67, 61, 71, 71, 71, 74, 71, 67,
          69, 68, 68, 70, 63, 70, 58, 66, 65, 69,
          62, 75, 72, 67, 67, 67, 67, 73, 64, 58,
          72, 71, 62, 71, 67, 63, 65, 69, 61, 71,
          69, 69, 70, 63, 69, 65, 66, 68, 72, 60,
          65, 66, 67, 64, 64, 70, 62, 65, 68, 67,
          66, 64, 68, 69, 67, 69, 60, 66, 73, 64,
          62, 75, 66, 65, 65, 68, 68, 71, 72, 70,
          67, 72, 73, 62, 68, 67, 75, 72, 74, 66, 
          68, 62, 67, 64, 60, 69, 70, 59, 65, 69]
sum(result) / 100.

67.17

The smartcab doesn't not always make to the destination. I ran the simulation 100 times, each time had a total trial of 100. On average, the car could make it at 67.17% of the time. Since all the moves were made randomly, the movement of the car was not making any sense. Sometimes it was hitting other cars, sometimes it just stopped at a green light.

### QUESTION 2
What states have you identified that are appropriate for modeling the smartcab and environment? Why do you believe each of these states to be appropriate for this problem?

To best make a decision, the following state variables should be considered,

+ the next waypoint: this gives us the general direction of where the destination is. We need to approach this direction to finish the trip. The values can be *None, forward, right, left*.

+ traffic light: this guides us where to go next to avoid punishment or/and get a reward. The values can be *red or green*.

+ presence of another vehicle at the intersection: since we are following Right-of-Way rules, we can then split the situation into four states:
    + if I want to go straight, I only care about the traffic light;
    + if I want to turn left, when the traffic light is red, is not allowed;
    + if I want to turn left, when the traffic light is green, I only care about oncoming cars which are going straight;
    + if I want to turn right, when the traffic light is red, I only care about cars on the left;
    + if I want to turn right, when the traffic light is green, I have the right of the way, so I don't care about other cars.

Given the above thoughts, the values we need to consider are, oncoming traffic - yes or not, and traffic on my left - yes or no.

+ deadline: we care about time left. Depending on what rewards we can get from taking each action, if the reward of getting to destination is really high, deadline can be vital since our only goal is to maximize our rewards. An extreme case could be we prefer breaking the traffic law, if our destination reward is so high ( a real life example could be a cop chasing a criminal).

### QUESTION 2 - OPTIONAL
How many states in total exist for the smartcab in this environment? Does this number seem reasonable given that the goal of Q-Learning is to learn and make informed decisions about each state? Why or why not?

There are 32 states in total. This number seems reasonable to me, as it is fairly small number of states, calculating it would not be difficult.